### ISSUE TITLE
Virtual batch norm, independent of number of replicas

### ISSUE THREAD

**Issue by znado** [google/flax#332]
### Description of the model to be implemented
Virtual batch normalization is useful to achieve SOTA performance in several benchmarks. However, in the distributed setting, the implementation must change how it computes virtual batches based on whether or not the per-replica batch size is less than or greater than the virtual batch size. For example, if the per-replica batch size is 32, but the virtual batch size is 128, then virtual batch norm should reduce across 4 replicas to compute the virtual batch. However, if the per-core batch size is 256, but the virtual batch size is 128, then virtual batch norm should split the per-core batch into 2 virtual batches.

Additionally, each virtual batch may need to maintain its own EMA for use at eval time. Some implementations such as [1] will calculate EMAs using a virtual batch and then average across all virtual batch EMAs before evaluating, while others such as [2] average over virtual batches before computing a per-core EMA (and, depending on the training library, may only evaluate with the EMA from a single replica). Supporting only the behavior in [1] would be sufficient, but having a parameter to switch between [1] and [2] may be more useful. Note that calculating virtual batch EMAs in the style of [1] would require maintaining multiple EMAs per replica, when in the setting where the per-core batch size is greater than the virtual batch size (in the example above, each replica would need to maintain two EMAs calculated with a virtual batch size of 128, and average over the two and across all replicas before evaluating).

### Dataset the model could be trained on
ImageNet/ResNet-50, as in [1]

### Reference implementations in other frameworks
[1] https://github.com/mlperf/training_results_v0.6/blob/master/Google/benchmarks/resnet/implementations/tpu-v3-2048-resnet/resnet/resnet_model.py#L151

[2] https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/normalization.py#L794

**Comment by jheek** [comment]
Thanks, this looks life a useful feature to have. I added PR welcome if somebody wants to give it a shot

**Comment by marcvanzee** [comment]
@jheek given that we didn't do this in 2 years, do you think it is worth keeping the issue, or shall we close it due to inactivity?

**Comment by marcvanzee** [comment]
@jheek assigned it to you so you can make a decision, feel free to unassign yourself once you decided what to do with it!

**Comment by jheek** [comment]
Nowadays JAX has a lot more tools to do this. You can define a model with a virtual_batch_size and "schedule" it using pjit or xmap anyway you want. I don't think we need custom support for this in Flax anymore
